### PR TITLE
Move the handling of labels into code generation.

### PR DESCRIPTION
**Pull request by gafter**
The handling of labels is significantly simplified and moved to code generation exclusively.

This is built on top of #43, so it will have to be merged and retargeted to `main` before merging.

### CODE REVIEW

File: src/Match2Cases.jl [hunk 1]
```diff
@@ -87,10 +87,15 @@ function generate_code(top_down_states::Vector{DeduplicatedCodePoint}, value, lo
...
+    labels = IdDict{DeduplicatedCodePoint, Symbol}()
+    function label(code::DeduplicatedCodePoint)
+        get!(() -> gensym("label", state), labels, code)
+    end
+
     while !isempty(togen)
         pc = pop!(togen)
-        if pc.label isa Symbol
-            push!(emit, :(@label $(pc.label)))
+        if pc in keys(labels)
```
Comment (mcmcgrath13) [inline src/Match2Cases.jl:97]: nifty!

this works because we are guaranteed to emit the `@goto` before we hit the node where we need the `@label`, right?

Comment (mcmcgrath13) [reply src/Match2Cases.jl:97]: if so, could you add a comment to that effect maybe before line 90?

File: src/StateMachine.jl [hunk 2]
```diff
@@ -305,11 +293,7 @@ function deduplicate_state_machine(entry::CodePoint, state::BinderState)
     top_down_states = reachable_states(entry)
     bottom_up_states = Iterators.reverse(top_down_states)
     for e in bottom_up_states
-        d = dedup(dedup_map, e, state)
-        if d.label === nothing
-            # It is a newly seen state
-            push!(result, d)
-        end
+        dedup(dedup_map, e, state)
```
Comment (mcmcgrath13) [inline src/StateMachine.jl:296]: should `dedup` be `dedup!` - it looks like we are relying on mutation here as the main effect. It also looks like `result` isn't used anymore

File: test/coverage.jl [hunk 3]
```diff
@@ -13,7 +13,11 @@ struct Trash <: Rematch2.BoundFetchPattern
     location::LineNumberNode
 end
 
-expected="""
+@testset "Tests that add code coverage" begin
+
+@testset "Check the actual output of dumpall" begin
```
Comment (mcmcgrath13) [inline test/coverage.jl:18]: indent the inner testsets for readability? (or at least comment the `end`s with what they match to)

**Review by mcmcgrath13 — APPROVED** [review]

File: src/Match2Cases.jl (hunk 1 above)
Comment (gafter) [reply src/Match2Cases.jl:97]: Yes.  That is because the nodes are topologically sorted.

**Pull request merged**